aten/src/ATen/BatchedTensorImpl.cpp

```diff
@@ -0,0 +1,104 @@
+#include <ATen/BatchedTensorImpl.h>
+
+#include <ATen/WrapDimUtils.h>
+#include <c10/util/Exception.h>
+
+namespace at {
+
+BatchedTensorImpl::BatchedTensorImpl(Tensor value, BatchDims bdims)
+  : TensorImpl(
+      c10::DispatchKeySet(DispatchKey::Vmap),
+      value.dtype(),
+      value.device()
+    )
+  , value_(std::move(value))
+  , bdims_(std::move(bdims))
+{
+  TORCH_INTERNAL_ASSERT(value_.defined());
+
+  const auto public_dims = value_.dim() - bdims_.size();
+  const auto value_sizes = value_.sizes();
+  sizes_.clear();
+  sizes_.reserve(public_dims);
+  for (int64_t dim = 0; dim < public_dims; dim++) {
+    auto actual_dim = actualDim(dim, /*wrap_dim=*/false);
+    sizes_.push_back(value_sizes.at(actual_dim));
+  }
+  refresh_numel();
+}
+
+int64_t BatchedTensorImpl::actualDim(int64_t dim, bool wrap_dim) const {
+  if (wrap_dim) {
+    const auto ndim = sizes_.size();
+    dim = maybe_wrap_dim(dim, ndim);
+  }
+  auto is_bdim = createBatchDimBitset(bdims_);
+
+  // Example: assume dim = 3, and is_bdim = 10010011000...
+  // The 1's are batch dims and 0's are normal dims of the underlying value_ Tensor.
+  // actualDim gives us the index of `dim` in the `value_` Tensor, which is equivalent
+  // to asking "where does the 3rd (0-indexed) zero occur in the bitset?".
+  // The answer to that is index 5.
+  int64_t non_bdim_count = 0;
+  for (int64_t actual_dim = 0; actual_dim < kVmapMaxTensorDims; actual_dim++) {
+    if (is_bdim[actual_dim]) {
+      continue;
+    }
+    if (non_bdim_count == dim) {
+      return actual_dim;
+    }
+    non_bdim_count++;
+  }
+  // If we hit this assert, then that means
+  // `non_bdim_count` + #num_bdims > kVmapMaxTensorDims. We restrict the number
+  // of dims a BatchedTensorImpl can have to kVmapMaxTensorDims so this should
+  // never be hit.
+  TORCH_INTERNAL_ASSERT(false);
+}
+
+// The following are publically exposed as methods of Tensor
+IntArrayRef BatchedTensorImpl::strides() const {
+  TORCH_CHECK(false, "NYI: Getting tensor strides inside of vmap");
```

Comment:
Arguably you can give meaningful strides for a batch tensor. The idea is that you give the strides that, if correspondingly interpreted, would let you address the FIRST batch in the batched tensor impl (so the strides of the public dimensions, not the private one).

Reply:
I agree that there is an argument that you can give meaningful strides. The reason why I am hesitant to provide it is because when users ask for strides, they are doing one of the following:
- they're about to grab the .data_ptr() and do some low-level things. vmap in its current state can't really support that.
- they're using the stride information to do some "performance tricks"; e.g. if you know what the strides of two matrices A, B are, it might be more performant to multiply them in some order (AB vs B^T A^T). The way vmap reorders dimensions could invalidate said performance tricks